library/common-library/templates/_cluster.tpl

```diff
@@ -2,12 +2,15 @@
 Return the cluster
 */}}
 {{- define "common.cluster" -}}
-    {{- if .Values.cluster -}}
-        {{- .Values.cluster -}}
-    {{- else if .Values.global -}}
-        {{- if .Values.global.cluster -}}
-            {{- .Values.global.cluster -}}
-        {{- end -}}
-    {{- end -}}
+{{- /* This allows us to use `$global` as an empty dict directly in case `Values.global` does not exists */ -}}
+{{- $global := index .Values "global" | default dict -}}
+
+{{- if .Values.cluster -}}
+    {{- .Values.cluster -}}
+{{- else if $global.cluster -}}
+    {{- $global.cluster -}}
+{{- else -}}
+    {{ fail "There is not cluster name definition set neither in `.global.cluster' nor `.cluster' in your values.yaml. Cluster name is required." }}
```

Comment:
Not a huge fan of templates that `fail` (I did not like that about the licenseKey helper either).

Rather than this logic that is hard to override, why wouldn't we just:

```
clusterName: {{ include "common.cluster" . | required "You must supply a clusterName" }}
```

Whenever we need it?

Reply:
We are not error-handling lots of things because we think that the user is clever enough to create good objects like in `affinity, `tolerations` or `hostNetwork: "SureBro"` instead of a boolean. But in that case we simply template it and Kubernetes will yell at the user. In this case is a dependency from us so it has to fail.

Here we have two approaches:
 - I trust what the common-library does and simply use it. I do not neither test it not ask myself what it does under the hood. I simply use it. So the common library has to do also the error handling.
 - I do not trust the common-library and retest what comes from it.

I think the first approach is the best because it *always* has to fail when no cluster name is provided so it leads to code repetition through all the charts.

Nevertheless, [In my last commit](https://github.com/newrelic/helm-charts/pull/759/commits/2304dc8cea873e705ee767b591bded35e5d013bc) I added a function to [change that behavior](https://github.com/newrelic/helm-charts/pull/759/commits/2304dc8cea873e705ee767b591bded35e5d013bc) if needed.

Reply:
I am fine with .1 and with having the fail explicit in the library.

(as always I would avoid having already the function to change the behaviour if we do not need it already)

Reply:
We have a couple of defaults that now I see that we don't need and other functions that I did not think that I was going to need and are handy to use (like the `.value` ones).

Avoiding having the function means that when you need it you will have no access to it and you'll have to develop it, review, merge and update and test all the charts.

Having a default function is like the "de facto" style that we have everywhere and It does no harm.

(I also think you are right)

Reply:
Like this
![image](https://c.tenor.com/DD_uRNFuG1YAAAAC/monkey-pointing.gif)

Reply:
We have a deal.

Reply:
I am going to leave this conversation unresolved so everybody can see it when doomscrolling.